### PR TITLE
[mqtt.homeassistant] support non-RGB lights

### PR DESCRIPTION
**Pull request by ccutrer**
because there's no rgbTopic, it was showing up as a trigger channel. which was super confusing, and commands were ignored because they are sent to the rgbTopic, which doesn't exist. so dynamically decide which type of channel to expose. also, there's still a pending bug where an RGB light will send ON/OFF and Brightness commands to the RGB command topic, instead of the specific type (if it exists).

### CODE REVIEW

**Comment by ccutrer** [comment]
@lolodomo @antroids: I'm still working on this (need to do JSON schema; that commit is basically empty, and then a few other smaller cleanups), but those are additions, not refining what is there, so I'd be happy to get feedback earlier in the process. The biggest concern I have is how I'm adding the interface to the Jinja bundle - would it be better to add that interface to the core transform bundle (and then not need to change the Jinja dependency to `provide`)?

**Comment by ccutrer** [comment]
> Adding this interface to the transform bungle looks reasonable for me, but I'm not sure for what exact purpose TransformationService was designed. Anyway it should be approved by core-maintainers. Could you create appropriate PR.

https://github.com/openhab/openhab-core/pull/3091

**Comment by ccutrer** [comment]
@antroids: I've reduced this to just the single commit for default schema lights in order to get that through. I'll do template schema and JSON schema as separate PRs (and JSON first), since the jinja thing is turning out to be non-trivial to get worked out.

**Comment by antroids** [comment]
> @antroids: I've reduced this to just the single commit for default schema lights in order to get that through. I'll do template schema and JSON schema as separate PRs (and JSON first), since the jinja thing is turning out to be non-trivial to get worked out.

Thanks! 
Looks like the only reason TransformationServiceProvider was created is transformation of the HA state topics. 
Probably, the best way is to move out Jinjava into core feature, as mentioned in corresponding PR, and use it as dependency in HA addon, without Transformation service. TransformationServiceProvider can be deprecated.

**Comment by antroids** [comment]
The worst thing in the HA Light API is that Color Mode can be changed dynamically. 
The Color Mode should be set to the last received valid color or updated by color_mode_state_topic, if specified.
Still not clear for me what should we do if Color Mode was updated from color_mode_state_topic, but the last received color state is in different mode.

> Home Assistant internally assumes that a light’s state corresponds to a defined color_mode. The state of MQTT lights with default schema and support for both color and color temperature will set the color_mode according to the last received valid color or color temperature. Optionally, a color_mode_state_topic can be configured for explicit control of the color_mode

https://github.com/home-assistant/core/blob/dev/tests/components/mqtt/test_light.py

**Comment by ccutrer** [comment]
> The worst thing in the HA Light API is that Color Mode can be changed dynamically. The Color Mode should be set to the last received valid color or updated by color_mode_state_topic, if specified. Still not clear for me what should we do if Color Mode was updated from color_mode_state_topic, but the last received color state is in different mode.
> 
> > Home Assistant internally assumes that a light’s state corresponds to a defined color_mode. The state of MQTT lights with default schema and support for both color and color temperature will set the color_mode according to the last received valid color or color temperature. Optionally, a color_mode_state_topic can be configured for explicit control of the color_mode
> 
> https://github.com/home-assistant/core/blob/dev/tests/components/mqtt/test_light.py

Yup. I plan on addressing color_mode if/when I implement color_temperature. I have some RGB+CCT lights, so I understand the necessity and confusion of changing color modes, but they're not currently exposed via Home Assistant compatible schema, so I can't directly use them to test. As for color_mode_state_topic, I see that for if you're currently in RGB mode, and someone presses "white" on an external controller - HASS/OpenHAB should then know that the light is now on the last known state for white color mode (brightness if it only supports brightness, color temp if it's CCT, etc.). I was looking through HASS sources the other day, and was surprised to find that HASS's RGBW and RGBWW modes support a light that can do RGB _and_ white at the same time. My RGB+CCT are either in RGB mode, or CCT mode. But HASS has code to support using white for the base level for the brightness, and a minimum amount of RGB in order to color it correctly. I thought that was interesting (and will be one of the last things implemented, if ever, for this binding - cause complicated math is involved when it's CCT and not just RGBW).

Anyhow, just making sure I read your comments correctly - you don't currently have any specific feedback for this PR, right? Just some comments for future directions. My personal priority at present is supporting on/off and brightness only lights, and then template schema, then effect, since those are what the light I'm trying to integrate use/have (it's an RGB light, but is not fully controllable - it's on/off, and you can set a limited number of named colors, or some effects that make it automatically transition among colors).

**Comment by antroids** [comment]
> Anyhow, just making sure I read your comments correctly - you don't currently have any specific feedback for this PR, right? Just some comments for future directions.

Yes, you are right. The full API support is out of scope of this PR.

**Comment by ccutrer** [comment]
@antroids : are you good with this PR now, or do you still have more to review?

**Review by antroids — CHANGES_REQUESTED** [review]
Great work with this overcomplicated API.
I found only one possible NPE that should be reviewed and several cosmetic issues.
Unfortunately I cannot test it on appropriate hardware.

UPD: some of the issues can be found in the SAT report at target/code-analysis/report.html
It will be generated after execution of maven verify stage.

File: bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/component/DefaultSchemaLight.java [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+    }
+
+    @Override
+    protected void buildChannels() {
+        ComponentChannel localOnOffChannel;
+        localOnOffChannel = onOffChannel = buildChannel(ON_OFF_CHANNEL_ID, onOffValue, "On/Off State", this)
+                .stateTopic(channelConfiguration.stateTopic, channelConfiguration.stateValueTemplate)
+                .commandTopic(channelConfiguration.commandTopic, channelConfiguration.isRetain(),
+                        channelConfiguration.getQos())
+                .commandFilter(command -> handleRawOnOffCommand(command)).build(false);
```
[the remark this answers is not in the document]
Comment (ccutrer) [reply bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/component/DefaultSchemaLight.java]: thanks. one of those java-isms that slowly I'm getting better at

File: bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/component/Light.java [hunk 2]
```diff
@@ -12,6 +12,8 @@
  */
 package org.openhab.binding.mqtt.homeassistant.internal.component;
 
+import java.math.BigDecimal;
+import java.util.ArrayList;
 import java.util.List;
```
Comment (antroids) [inline bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/component/Light.java:17]:
```suggestion
import java.util.List;
import java.util.Objects;
```

**Review by antroids — CHANGES_REQUESTED** [review]
Let's make these lambdas uniform

**Review by lolodomo — APPROVED** [review]
LGTM, thank you

**Pull request merged**